internal/controller/bmc_controller.go

```diff
@@ -193,10 +244,255 @@ func (r *BMCReconciler) discoverServers(ctx context.Context, log logr.Logger, bm
+				if err := r.updateConditions(ctx, bmcObj, true, bmcReadyConditionType, corev1.ConditionFalse, bmcInternalErrorReason, "BMC internal server error"); err != nil {
+					return fmt.Errorf("failed to set BMC internal server error condition: %w", err)
+				}
+			}
+		case 503:
+			{
+				// Service Unavailable, might be transient
+				if err := r.updateConditions(ctx, bmcObj, true, bmcReadyConditionType, corev1.ConditionFalse, bmcConnectionFailedReason, "BMC service unavailable"); err != nil {
+					return fmt.Errorf("failed to set BMC service unavailable condition: %w", err)
+				}
+			}
+		default:
+			{
+				if err := r.updateConditions(ctx, bmcObj, true, bmcReadyConditionType, corev1.ConditionFalse, bmcUnknownErrorReason, fmt.Sprintf("BMC connection error: %v", err)); err != nil {
+					return fmt.Errorf("failed to set BMC error condition: %w", err)
+				}
+			}
+		}
+	} else {
+		if err := r.updateConditions(ctx, bmcObj, true, bmcReadyConditionType, corev1.ConditionFalse, bmcUnknownErrorReason, fmt.Sprintf("BMC connection error: %v", err)); err != nil {
+			return fmt.Errorf("failed to set BMC error condition: %w", err)
+		}
+	}
+	return err
+}
 
+func (r *BMCReconciler) shouldSkipReconciliation(bmcObj *metalv1alpha1.BMC) bool {
+	acc := conditionutils.NewAccessor(conditionutils.AccessorOptions{})
+	condition := &metav1.Condition{}
+	found, err := acc.FindSlice(bmcObj.Status.Conditions, bmcResetConditionType, condition)
+	if err != nil || !found {
+		return false
+	}
+	if condition.Status == metav1.ConditionTrue {
+		// give bmc some time to start the reset process
+		if time.Since(condition.LastTransitionTime.Time) < 5*time.Second {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *BMCReconciler) shouldResetBMC(bmcObj *metalv1alpha1.BMC) bool {
+	if r.BMCFailureResetDelay == 0 {
+		return false
+	}
+	acc := conditionutils.NewAccessor(conditionutils.AccessorOptions{})
+	bmcResetCondition := &metav1.Condition{}
+	found, err := acc.FindSlice(bmcObj.Status.Conditions, bmcResetConditionType, bmcResetCondition)
+	if err != nil || (found && bmcResetCondition.Status == metav1.ConditionTrue) {
+		return false
+	}
+	readyCondition := &metav1.Condition{}
+	found, err = acc.FindSlice(bmcObj.Status.Conditions, bmcReadyConditionType, readyCondition)
+	if err != nil || !found {
+		return false
+	}
+	if readyCondition.Status == metav1.ConditionFalse && (readyCondition.Reason == bmcInternalErrorReason || readyCondition.Reason == bmcConnectionFailedReason) {
+		if time.Since(readyCondition.LastTransitionTime.Time) > r.BMCFailureResetDelay {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *BMCReconciler) resetBMC(ctx context.Context, log logr.Logger, bmcObj *metalv1alpha1.BMC, reason, message string) error {
+	running, err := r.checkResetJobStatus(ctx, log, bmcObj)
+	if err != nil {
+		return fmt.Errorf("BMC reset job status: %w", err)
+	}
+	if running {
+		// job is still running, requeue
+		return fmt.Errorf("BMC reset job is still running")
+	}
```

Comment:
`checkResetJobStatus()` returns an error for various state of Job, except when the job is running. Could the check for running be moved into `checkResetJobStatus()` as well?